File: src/main/java/com/telenordigital/prime/ocs/OcsService.java
package com.telenordigital.prime.ocs;

import com.lmax.disruptor.EventHandler;
import com.telenordigital.prime.disruptor.PrimeEvent;
import com.telenordigital.prime.disruptor.PrimeEventProducer;
import io.grpc.stub.StreamObserver;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

import static com.google.common.base.Preconditions.checkNotNull;

/**
 * @author Vihang Patil <vihang.patil@telenordigital.com>
 */
public final class OcsService  {

    private static final Logger LOG = LoggerFactory.getLogger(OcsService.class);

    private final  ConcurrentMap<String, StreamObserver<FetchDataBucketInfo>>
            fetchDataBucketClientMap;

    private final ConcurrentMap<String, StreamObserver<ReturnUnusedDataResponse>>
            returnUnusedDataClientMap;

    private ActivateResponseHolder activateResponseHolder;

    private final PrimeEventProducer producer;

    private final EventHandler<PrimeEvent> eventHandler;

    private final OcsServiceGrpc.OcsServiceImplBase ocsServerImplBaseImpl;

    public OcsService(final PrimeEventProducer producer) {
        this.producer = checkNotNull(producer);
        this.fetchDataBucketClientMap = new ConcurrentHashMap<>();
        this.returnUnusedDataClientMap = new ConcurrentHashMap<>();
        this.eventHandler = new EventHandlerImpl(this);
        this.ocsServerImplBaseImpl = new OcsServerImplBaseImpl(this );
        this.activateResponseHolder = new ActivateResponseHolder();
    }

    public EventHandler<PrimeEvent> asEventHandler() {
        return eventHandler;
    }

    protected void returnUnusedDataBucketEvent(final String msisdn, final long bucketBytes) {
        producer.returnUnusedDataBucketEvent(
                msisdn,
                bucketBytes,
                null);
    }

    public OcsServiceGrpc.OcsServiceImplBase asOcsServiceImplBase() {
        return this.ocsServerImplBaseImpl;
    }

    protected StreamObserver<FetchDataBucketInfo> getDataBucketClientForStream(final String streamId) {
         return fetchDataBucketClientMap.get(streamId);
    }

    protected StreamObserver<ReturnUnusedDataResponse> getUnusedDataClientForStream(final String streamId) {
         return returnUnusedDataClientMap.get(streamId);
    }

    protected void activateOnNextResponse(final ActivateResponse response) {
        this.activateResponseHolder.onNextResponse(response);
    }

    protected void updateActivateResponse(final StreamObserver<ActivateResponse> activateResponse) {
        this.activateResponseHolder.setActivateResponse(activateResponse);
    }

    protected void removeUnusedDataClient(String streamId) {
        this.returnUnusedDataClientMap.remove(streamId);
    }

    protected void returnUnusedDataBucketEvent(
            final String msisdn,
            final long bytes,
            final String streamId) {
        this.producer.returnUnusedDataBucketEvent(
                msisdn, bytes, streamId);
    }

    protected void registerUnusedDataClient(
            final String streamId,
            final StreamObserver<ReturnUnusedDataResponse> returnUnusedDataResponse) {
        this.returnUnusedDataClientMap.put(streamId, returnUnusedDataResponse);
    }

    protected void deleteDataBucketClient(final String streamId) {
        this.fetchDataBucketClientMap.remove(streamId);
    }

    protected  void fetchDataBucketEvent(
            final FetchDataBucketInfo request,
            final String streamId) {
        producer.fetchDataBucketEvent(request, streamId);
    }

    protected void putDataBucketClient(
            final String streamId,
            final StreamObserver<FetchDataBucketInfo> fetchDataBucketResponse) {
        fetchDataBucketClientMap.put(streamId, fetchDataBucketResponse);
    }
}

File: ocs/src/main/kotlin/org/ostelco/prime/consumption/OcsService.kt
package org.ostelco.prime.consumption

import com.lmax.disruptor.EventHandler
import io.grpc.stub.StreamObserver
import org.ostelco.ocs.api.ActivateResponse
import org.ostelco.ocs.api.CreditControlAnswerInfo
import org.ostelco.ocs.api.CreditControlRequestInfo
import org.ostelco.ocs.api.OcsServiceGrpc
import org.ostelco.prime.disruptor.EventProducer
import org.ostelco.prime.disruptor.OcsEvent
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentMap

/**
 * [OcsService] acts as bridge between [OcsGrpcService] and [OcsDisruptor].
 *
 * Ocs Requests from [OcsGrpcService] are sent to [OcsService] via [OcsAsyncRequestConsumer] interface.
 * These requests are then sent to [OcsDisruptor] via [EventProducer].
 * [EventProducerImpl] is the implementation for [EventProducer] interface.
 *
 * OcsEvents from [OcsDisruptor] are handled by [OcsEventToGrpcResponseMapper].
 * [OcsEventToGrpcResponseMapper] forwards responses to [OcsService] via [OcsAsyncResponseProducer] interface.
 * [OcsService] produces responses and sends them to [OcsGrpcService].
 */
class OcsService(private val producer: EventProducer) : OcsAsyncRequestConsumer, OcsAsyncResponseProducer {

    private val creditControlClientMap: ConcurrentMap<String, StreamObserver<CreditControlAnswerInfo>> = ConcurrentHashMap()
    private val activateMsisdnClientMap: ConcurrentMap<String, StreamObserver<ActivateResponse>> = ConcurrentHashMap()

    /**
     * A holder for
     * [<]
     * instances that are somehow used
     */
    val eventHandler: EventHandler<OcsEvent> = OcsEventToGrpcResponseMapper(this)
    val ocsGrpcService: OcsServiceGrpc.OcsServiceImplBase = OcsGrpcService(this)

    //
    // Request Consumer functions
    //

    override fun updateActivateResponse(
            streamId: String,
            activateResponse: StreamObserver<ActivateResponse>) {
        this.activateMsisdnClientMap[streamId] = activateResponse
    }

    override fun deleteCreditControlClient(streamId: String) {
        this.creditControlClientMap.remove(streamId)
    }

    override fun creditControlRequestEvent(
            streamId: String,
            request: CreditControlRequestInfo) {
        producer.injectCreditControlRequestIntoRingbuffer(streamId, request)
    }

    override fun putCreditControlClient(
            streamId: String,
            creditControlAnswer: StreamObserver<CreditControlAnswerInfo>) {
        creditControlClientMap[streamId] = creditControlAnswer
    }

    //
    // Response Producer functions
    //

    override fun sendCreditControlAnswer(
            streamId: String,
            creditControlAnswer: CreditControlAnswerInfo) {
        creditControlClientMap[streamId]?.onNext(creditControlAnswer)
    }

    override fun activateOnNextResponse(response: ActivateResponse) {
        // TODO martin: send activate MSISDN to selective ocsgw instead of all
        this.activateMsisdnClientMap.forEach { _ , responseStream ->
            responseStream.onNext(response)
        }
    }

    override fun returnUnusedDataBucketEvent(
            msisdn: String,
            reservedBucketBytes: Long) {
        producer.releaseReservedDataBucketEvent(
                msisdn,
                reservedBucketBytes)
    }
}
